lib/erb_lint/linters/final_newline.rb

```diff
@@ -10,26 +10,33 @@ def initialize(config)
         @new_lines_should_be_present = config['present'].nil? ? true : config['present']
       end
 
-      protected
-
-      def lint_lines(lines)
-        errors = []
-        return errors if lines.empty?
-
-        ends_with_newline = lines.last.chars[-1] == "\n"
-
-        if @new_lines_should_be_present && !ends_with_newline
-          errors.push(
-            line: lines.length,
-            message: 'Missing a trailing newline at the end of the file.'
-          )
-        elsif !@new_lines_should_be_present && ends_with_newline
-          errors.push(
-            line: lines.length,
-            message: 'Remove the trailing newline at the end of the file.'
-          )
+      def lint_file(file_tree)
```

Comment:
@lemonmade this comment got unintentionally marked as resolved by Gtihub.
>Is there no access to the source of the file? You're jumping through a lot of hoops to make this work with the parsed tree.

The idea behind using a parse tree always is that it provides a consistent interface for all linters to use and that most future linters (data-attributes, product content style, etc) will be complex enough to leverage the benefits a tree brings.

The reason that I'm jumping through the hoops here in `FinalNewline` is only because of the [Nokogiri line number bug](https://github.com/sparklemotion/nokogiri/issues/1493). Once that is fixed, the hacky `end_marker` will be completely removed and this linter will be much simpler, just finding the last child and calling `#line`.

As a compromise would you suggest a wrapper class around both the Parse tree and a copy of the original source?